### PR TITLE
Adding Token Counter for Online RL

### PR DESCRIPTION
**Pull request by rithwik-db**
We are updating the token counter to either be defined by the `action_mask` + `prompt_len` or using the `len(sequences)` when pad token is removed. I tested it out on `compose-rl-grpo-test-5w6ZMR` and both return the same output.

### CODE REVIEW

File: compose_rl/data/dataloader.py [hunk 1]
```diff
@@ -102,7 +105,10 @@ def build_preference_dataloader(
             persistent_workers=persistent_workers,
             timeout=timeout,
         )
-        return dataloader
+        data_spec_kwargs = {}
+        if token_counter_fn is not None:
+            data_spec_kwargs['get_num_tokens_in_batch'] = partial(token_counter_fn, pad_token_id=tokenizer.pad_token_id)
```
Comment (rithwik-db) [inline compose_rl/data/dataloader.py:110]: In theory, we can remove the `pad_token_id` if we just stick with the `action_mask + prompt_len` but was curious to hear what is preferred.

**Comment by bowenyang008** [comment]
Thanks @rithwik-db it looks good to me but I would like to have reviews from experts. Additionally I found it pretty useful to log:
```
[RICKY] prompt_tokens: 1701
[RICKY] generated_tokens: 463
```
per iteration or per minibatch in each iteration. If no objection from other reviewers I would prefer adding this to the log as you have already done.

**Comment by bowenyang008** [comment]
also comparing your log and one of my log on main grpo-t2s-lr-1e-6-clip-5e-3-kl-1e-3-v7-9JJUW2

there is an order of magnitude diff
this branch:
Train throughput/tokens_per_sec: 12892.0938
main:
Train throughput/tokens_per_sec: 106238.8025
and the MFU using this branch is only 8%

**Comment by bowenyang008** [comment]
I also benchmarked the dataset via this run: grpo-t2s-lr-1e-6-clip-5e-3-kl-1e-3-v7-2D7L30 and got this:

```
2025-07-16 00:29:58,782: rank0[820][MainThread]: INFO: compose_rl.algorithms.online.callback: number of prompts in full train_prompt_loader dataset: 9432
2025-07-16 00:29:59,349: rank0[820][MainThread]: INFO: compose_rl.algorithms.online.callback: global max prompt length: 3212
2025-07-16 00:29:59,349: rank0[820][MainThread]: INFO: compose_rl.algorithms.online.callback: global mean prompt length: 1641.726887192536
```

**Comment by rithwik-db** [comment]
@bowenyang008, added the logs you mentioned [here](https://github.com/databricks/compose-rl/pull/110#issuecomment-3076232144).

To answer:

> so most the tokens are just padding?

Not exactly, the max amount of generated tokens is 2000 (which is what `max_gen_len` is) in the yaml above and so the amount of padding can't exceed that value. It's just that when we count tokens, we use `Composer`'s counter which ends up just being `max_seq_len * num_samples_in_batch` which is defined [here](https://github.com/mosaicml/composer/blob/ef6315f5863dcfcdf9bd484e5d4d62cdfaae46de/composer/core/data_spec.py#L297). Added this to the PR description as well.

File: compose_rl/data/dataloader.py [hunk 2]
```diff
@@ -117,9 +129,31 @@ def build_preference_dataloader(
...
+    """
+    if 'action_mask' in batch and 'prompt_len' in batch:
+        prompt_len_tokens = batch['prompt_len'].sum().item()
+        action_mask_tokens = batch['action_mask'].sum().item()
+        log.info(f'prompt tokens in batch: {prompt_len_tokens}')
+        log.info(f'generated tokens in batch: {action_mask_tokens}')
+        return int(prompt_len_tokens + action_mask_tokens)
+    elif 'sequences' in batch:
+        sequences = batch['sequences']
+        non_pad_mask_tokens = torch.ne(sequences, pad_token_id).sum().item()
```
Comment (bowenyang008) [inline compose_rl/data/dataloader.py:147]: can we also log padded tokens as well?

Comment (rithwik-db) [reply compose_rl/data/dataloader.py]: @bowenyang008 I added the logging for padded tokens but it's a bit odd since there's two different calculations (padding that we manually add vs padding that already exists because of unused generation capacity).

File: compose_rl/data/dataloader.py (hunk 1 above)
Comment (rithwik-db) [reply compose_rl/data/dataloader.py]: I think the `action_mask + prompt_len` would probably be the better route since `sequences` isn't really used outside of the callback whereas `action_mask` is used for training

File: compose_rl/data/dataloader.py [hunk 3]
```diff
@@ -117,9 +125,36 @@ def build_preference_dataloader(
...
+        log.info(
+            f'unused generation (padding) tokens in batch: {padding_tokens}',
+        )
+        return int(prompt_len_tokens + generated_items)
+    else:
+        log.warning(
+            'No action_mask/prompt_len in batch. ' +
+            'Using default value of 0 for num_tokens_in_batch.',
+        )
+        return 0
```
Comment (rithwik-db) [inline compose_rl/data/dataloader.py:151]: Using 0 since [composer default is also 0](https://sourcegraph.prod.databricks-corp.com/mosaicml/composer/-/blob/composer/core/data_spec.py?L306) and we don't want to raise an error here so I logged a warning instead.

File: compose_rl/algorithms/online/model_methods.py [hunk 4]
```diff
@@ -246,8 +246,9 @@ def policy_loss(
         token_entropies = utils.get_token_entropies(
             logits=gen_logits,
         )
-        assert token_entropies.shape == batch['action_mask'].shape, (
-            f'Token entropies shape {token_entropies.shape} does not match action mask shape {batch["action_mask"].shape}.',
+        action_mask = batch['action_mask']
```
Comment (gupta-abhay) [inline compose_rl/algorithms/online/model_methods.py:249]: are we defining this variable just for the check? If not, maybe best to remove or propagate the usage throughout the function scope ?

File: compose_rl/data/dataloader.py [hunk 5]
```diff
@@ -117,9 +125,36 @@ def build_preference_dataloader(
...
+
+
 build_prompt_dataloader = generate_dataloader_builder(
     PromptStreamingDataset,
     prompt_dataset_collate_fn,
+    get_num_tokens_in_batch_online,
```
Comment (gupta-abhay) [inline compose_rl/data/dataloader.py:157]: can we please add this functionality to the `messages_dataloader` as well?

**Review by gupta-abhay — COMMENTED** [review]
some comments

File: compose_rl/algorithms/online/model_methods.py (hunk 4 above)
Comment (rithwik-db) [reply compose_rl/algorithms/online/model_methods.py:249]: I am getting formatting errors without this change so I guess I will just propagate this

File: compose_rl/data/dataloader.py (hunk 5 above)
Comment (rithwik-db) [reply compose_rl/data/dataloader.py:157]: My understanding is that the same code should work for `messages_dataloader` as well, but if you have a yaml or something that I can use to test this out, would be much appreciated!
